src/control/positionControl/PositionControl.cpp

```diff
@@ -6,44 +6,215 @@
+                    if (worldObjects.canMoveOutsideField(robotId) ||
+                        rtt::ai::FieldComputations::pointIsInField(field, intermediatePoint,
+                                                                   rtt::ai::Constants::ROBOT_RADIUS())) {
+                        intermediatePoints.emplace_back(intermediatePoint);
+                    }
+                }
+            }
+        }
+        return intermediatePoints;
+    }
+
+    std::priority_queue<std::pair<double, Vector2>, std::vector<std::pair<double, Vector2>>, std::greater<>>
+    PositionControl::scoreIntermediatePoints(std::vector<Vector2> &intermediatePoints,
+                                            std::optional<BB::CollisionData> &firstCollision) {
+        double intermediatePointScore;
+        std::priority_queue<std::pair<double, Vector2>, std::vector<std::pair<double, Vector2>>, std::greater<>> intermediatePointsSorted;
+        for (auto i : intermediatePoints) {
+            intermediatePointScore = (i - firstCollision->collisionPosition).length();
+            std::pair<double, Vector2> p = {intermediatePointScore, i};
+            intermediatePointsSorted.push(p);
+        }
+        return intermediatePointsSorted;
+    }
+
+    std::optional<BB::BBTrajectory2D>
+    PositionControl::calculatePathFromNewStart(std::optional<BB::CollisionData> intermediatePathCollision,
+                                               BB::BBTrajectory2D pathToIntermediatePoint,
+                                               Vector2 &targetPosition, int robotId, double timeStep) {
+        BB::BBTrajectory2D intermediateToTarget;
+        if (!intermediatePathCollision.has_value()) {
+            timeStep *= 2;
+            for (int i = 0; i < floor(pathToIntermediatePoint.getTotalTime() / timeStep); i++) {
+                //TODO: Only create newStart's up to point where we dont decelerate yet
+                Vector2 newStart = pathToIntermediatePoint.getPosition(i * timeStep);
+                Vector2 newVelocity = pathToIntermediatePoint.getVelocity(i * timeStep);
 
-void PositionControl::setRobotPositions(std::vector<Vector2> &robotPositions) { collisionDetector.setRobotPositions(robotPositions); }
-}  // namespace rtt::ai::control
\ No newline at end of file
+                intermediateToTarget = BB::BBTrajectory2D(newStart, newVelocity, targetPosition,
+                                                          ai::Constants::MAX_VEL(),
+                                                          ai::Constants::MAX_ACC_UPPER());
+                auto newStartCollisions = worldObjects.getFirstCollision(intermediateToTarget, robotId);
+
+                if (newStartCollisions.has_value()) {
+                    continue;
+                } else {
+                    return intermediateToTarget;
+                }
+            }
+        }
+        return {};
```

Comment:
Is this returning nullopt or an empty trajectory? Can be very unclear, should probably make it more explicit.